tools/clang/lib/SPIRV/DeclResultIdMapper.cpp

```diff
@@ -1802,51 +1804,85 @@ class LocationSet {
-  const static uint32_t kMaxLoc = 64;
 
   LocationSet() {
     for (uint32_t i = 0; i < kMaxIndex; ++i) {
-      usedLocs[i].resize(kMaxLoc);
-      nextLoc[i] = 0;
+      // Default size. 64 should cover most cases without having to resize.
+      usedLocations[i].resize(64);
+      nextAvailableLocation[i] = 0;
     }
   }
 
-  /// Uses the given location.
+  /// Marks a given location as used.
   void useLoc(uint32_t loc, uint32_t index = 0) {
     assert(index < kMaxIndex);
-    usedLocs[index].set(loc);
-  }
 
-  /// Uses the next |count| available location.
-  int useNextLocs(uint32_t count, uint32_t index = 0) {
-    assert(index < kMaxIndex);
-    auto &locs = usedLocs[index];
-    auto &next = nextLoc[index];
-    while (locs[next])
-      next++;
+    auto &set = usedLocations[index];
+    if (loc <= set.size()) {
+      set.resize(std::max<size_t>(loc + 1, set.size() * 2));
+    }
+    set.set(loc);
 
-    int toUse = next;
+    // Maybe the bit marked is the next available one.
+    // Otherwise, there is a hole, ignoring.
+    if (loc == nextAvailableLocation[index]) {
+      ++nextAvailableLocation[index];
+    }
+  }
 
-    for (uint32_t i = 0; i < count; ++i) {
-      assert(!locs[next]);
-      locs.set(next++);
+  // Find the first range of size |count| of unused locations,
+  // and marks them as used.
+  // Returns the first index of this range.
+  int useNextLocs(uint32_t count, uint32_t index = 0) {
+    auto res = find_unused_range(index, count);
+    auto &locations = usedLocations[index];
+
+    // Simple case: no hole large enough left, resizing.
+    if (res == std::nullopt) {
+      res = locations.size();
+      locations.resize(
+          std::max<size_t>(locations.size() + count, locations.size() * 2));
     }
 
-    return toUse;
+    for (uint32_t i = res.value(); i < res.value() + count; i++) {
+      locations.set(i);
+    }
+    return res.value();
   }
 
   /// Returns true if the given location number is already used.
   bool isLocUsed(uint32_t loc, uint32_t index = 0) {
     assert(index < kMaxIndex);
-    return usedLocs[index][loc];
+    if (loc >= usedLocations[index].size())
+      return false;
+    return usedLocations[index][loc];
   }
 
 private:
-  llvm::SmallBitVector usedLocs[kMaxIndex]; ///< All previously used locations
-  uint32_t nextLoc[kMaxIndex];              ///< Next available location
+  // Find the first unused range of size |size| in the given set.
+  // If the set contains such range, returns the first usable index.
+  // Otherwise, nullopt is returned.
+  std::optional<uint32_t> find_unused_range(uint32_t index, uint32_t size) {
+    assert(index < kMaxIndex);
+    const auto &locations = usedLocations[index];
+    uint32_t slot_size = 0;
+    for (uint32_t i = nextAvailableLocation[index]; i < locations.size(); i++) {
+      if (locations[i])
+        slot_size = 0;
+      else
+        ++slot_size;
+
+      if (slot_size == size)
+        return i + 1 - slot_size;
+    }
```

Comment:
What happens if the size of `locations` is 1024 and is all empty and you want a range of size 1025?

Reply:
We bump the vector size to 1024+1025. Added a quick way to handle this specific case which doesn't add that much complexity. Would like to keep the logic simple as I believe those are corner cases. We were fine a fixed, unchecked 64 bits vector for quite some time, so don't believe we might hit those cases often.